src/driver/franka_plan_runner.cc

```diff
@@ -797,10 +868,237 @@ franka::JointPositions FrankaPlanRunner::JointPositionCallback(
+  Eigen::Map<const Eigen::Matrix<double, 6, 1>> wrench_limits {
+      wrench_limits_array.data()};
+
+  // Eigen::Matrix<double, 6, 1> task_wrench {(-stiffness * error -
+  // damping * (cart_vel))};
+  // // log()->info("task_wrench: {}", task_wrench.transpose());
+  // // task_wrench =
+  // task_wrench.cwiseMin(wrench_limits).cwiseMax(-wrench_limits);
+  // // log()->info("task_wrench*: {}", task_wrench.transpose());
+
+  Eigen::Matrix<double, 6, 1> task_wrench {
+      (-stiffness_ * error - damping_ * (cart_vel))};
+  task_wrench = task_wrench.cwiseMin(wrench_limits).cwiseMax(-wrench_limits);
+
+  // compute control
+  Eigen::Matrix<double, 7, 1> tau_task(7), tau_d(7), tau_joint_centering(7);
+
+  Eigen::Map<const Eigen::Matrix<double, 7, 1>> torque_limits(
+      kImpedanceControlTorqueThreshold.data());
+
+  // Spring damper system with damping ratio=1
+  tau_task << jacobian.transpose() * task_wrench;
+
+  const auto jc_spring {q_error * (-k_centering_)};
+  const auto jc_damping {dq * (-2 * sqrt(k_centering_))};
+
+  // 7x7 * 7x1
+  tau_joint_centering << /* _ * */ (jc_spring + jc_damping);
+  tau_joint_centering = tau_joint_centering * k_jc_ramp_;
+  tau_joint_centering =
+      tau_joint_centering.cwiseMin(torque_limits).cwiseMax(-torque_limits);
+
+  k_jc_ramp_ = k_jc_ramp_ * (1 - filter_gain_) + filter_gain_;
+
+  tau_d << tau_task + coriolis + tau_joint_centering;
+
+  // we print info in a separate thread to keep callback short
+  // TODO(@syler): demote verbosity or remove once tested
+  auto print_info {[tau_task, coriolis, tau_joint_centering, q_diff_from_center,
+                    q_diff_from_center_norm, q_error, jc_spring, jc_damping]() {
+    log()->info(
+        "\n\tTask:\t{}\n\tCoriolis:\t{}\n\tCentering:\t{}\n\tq_diff:\t{"
+        "}\n\tq_diff "
+        "norm:\t{}\n\tq_error:\t{}\n\tjc_spring:\t{}\n\tjc_damping:\t{"
+        "}",
+        tau_task.transpose(), coriolis.transpose(),
+        tau_joint_centering.transpose(), q_diff_from_center.transpose(),
+        q_diff_from_center_norm.transpose(), q_error.transpose(),
+        jc_spring.transpose(), jc_damping.transpose());
+  }};
+
+  std::thread print_info_thread {print_info};
+  print_info_thread.detach();
+
+  // torque saturation to limits
+  tau_d = tau_d.cwiseMin(torque_limits).cwiseMax(-torque_limits);
+
+  std::array<double, 7> tau_d_array {};
+  Eigen::Matrix<double, 7, 1>::Map(&tau_d_array[0], 7) = tau_d;
+
+  franka::Torques ret_torques {tau_d_array};
+  if (comm_interface_->CompliantPushStopRequested()) {
+    comm_interface_->ClearCompliantPushStopRequest();
+    ret_torques.motion_finished = true;
+    return ret_torques;
+  }
+
+  auto end_time {std::chrono::high_resolution_clock::now()};
+  time_elapsed_us_.push_back(
+      std::chrono::duration_cast<std::chrono::microseconds>(end_time
+                                                            - start_time)
+          .count());
```

Comment:
this is still 1khz right? why not use milliseconds as used everywhere else?

Reply:
Yes, but the loop takes about 120 us to run on average, and at abour 400 us, we start seeing communication constraint violations

Reply:
moved to separate thread